crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -131,46 +131,78 @@ impl CommandQueue {
-            // SAFETY: The address just past the command is either within the buffer,
-            // or 1 byte past the end, so this addition will not overflow the pointer's allocation.
-            cursor = unsafe { cursor.add(size) };
+        let bytes = std::mem::take(&mut self.bytes);
+
+        let mut resolving_commands = vec![(cursor, end, bytes)];
+
+        while let Some((mut cursor, mut end, mut bytes)) = resolving_commands.pop() {
+            while cursor < end {
+                // SAFETY: The cursor is either at the start of the buffer, or just after the previous command.
+                // Since we know that the cursor is in bounds, it must point to the start of a new command.
+                let meta = unsafe { cursor.cast::<CommandMeta>().read_unaligned() };
+                // Advance to the bytes just after `meta`, which represent a type-erased command.
+                // SAFETY: For most types of `Command`, the pointer immediately following the metadata
+                // is guaranteed to be in bounds. If the command is a zero-sized type (ZST), then the cursor
+                // might be 1 byte past the end of the buffer, which is safe.
+                cursor = unsafe { cursor.add(std::mem::size_of::<CommandMeta>()) };
+                // Construct an owned pointer to the command.
+                // SAFETY: It is safe to transfer ownership out of `self.bytes`, since the call to `set_len(0)` above
+                // guarantees that nothing stored in the buffer will get observed after this function ends.
+                // `cmd` points to a valid address of a stored command, so it must be non-null.
+                let cmd = unsafe {
+                    OwningPtr::<Unaligned>::new(std::ptr::NonNull::new_unchecked(cursor.cast()))
+                };
+                // SAFETY: The data underneath the cursor must correspond to the type erased in metadata,
+                // since they were stored next to each other by `.push()`.
+                // For ZSTs, the type doesn't matter as long as the pointer is non-null.
+                let size = unsafe { (meta.consume_command_and_get_size)(cmd, &mut world) };
+                // Advance the cursor past the command. For ZSTs, the cursor will not move.
+                // At this point, it will either point to the next `CommandMeta`,
+                // or the cursor will be out of bounds and the loop will end.
+                // SAFETY: The address just past the command is either within the buffer,
+                // or 1 byte past the end, so this addition will not overflow the pointer's allocation.
+                cursor = unsafe { cursor.add(size) };
+
+                if let Some(world) = &mut world {
+                    // Mark world as flushing so calls to `World::flush_commands` don't cause multiple queues to be applying simultaneously
+                    world.flushing_commands = true;
```

Comment:
What is the motivation adding this guard? Seems liked its only used in `World::flush_commands`, but you can get around it by calling `CommandQueue::apply` to flush a world's commands. And unless I'm mistaken, the guard isn't necessary either, as nested calls to `flush_commands` wouldn't introduce any inconsistent behavior.

Reply:
Yeah you're right, this was mangled when I merged the `apply_drop_or_queued` refactor. This logic isn't necessary any more as (especially after the change to ensure the world queue is flushed first) it doesn't prevent anything unintended.